spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/outbound/MqttPahoMessageHandler.java

```diff
@@ -154,41 +154,37 @@ protected void doStop() {
 		}
 	}
 
-	private synchronized void doConnect() throws MqttException {
+	private synchronized MqttAsyncClient checkConnection() throws MqttException {
 		if (this.client != null && !this.client.isConnected()) {
 			this.client.close();
 			this.client = null;
 		}
 		if (this.client == null) {
-			MqttConnectOptions connectionOptions = this.clientFactory.getConnectionOptions();
-			Assert.state(this.getUrl() != null || connectionOptions.getServerURIs() != null,
-					"If no 'url' provided, connectionOptions.getServerURIs() must not be null");
-			this.client = this.clientFactory.getAsyncClientInstance(this.getUrl(), this.getClientId());
-			incrementClientInstance();
-			this.client.setCallback(this);
-			this.client.connect(connectionOptions).waitForCompletion(this.completionTimeout);
-			if (logger.isDebugEnabled()) {
-				logger.debug("Client connected");
-			}
-		}
-	}
-
-	@Override
-	protected void connectIfNeeded() {
-		if (this.client == null || !this.client.isConnected()) {
 			try {
-				this.doConnect();
+				MqttConnectOptions connectionOptions = this.clientFactory.getConnectionOptions();
+				Assert.state(this.getUrl() != null || connectionOptions.getServerURIs() != null,
+						"If no 'url' provided, connectionOptions.getServerURIs() must not be null");
+				MqttAsyncClient client = this.clientFactory.getAsyncClientInstance(this.getUrl(), this.getClientId());
+				incrementClientInstance();
+				client.setCallback(this);
+				client.connect(connectionOptions).waitForCompletion(this.completionTimeout);
+				this.client = client;
+				if (logger.isDebugEnabled()) {
+					logger.debug("Client connected");
+				}
 			}
 			catch (MqttException e) {
 				throw new MessagingException("Failed to connect", e);
 			}
 		}
+		return this.client;
```

Comment:
I think for better protection against `null` we should return an internal variable and only after that assign it to the property.